### ISSUE TITLE
(PostgreSQL 13) 4 artifacts with user base types

### ISSUE THREAD

**Issue by vicreal** [dbeaver/dbeaver#10818]
**Windows 10 + DBeaver 7.2.4 + PostgreSQL 13** (tunnels or proxies not used).
The problems is only in the case of using **[custom base types](https://www.postgresql.org/docs/13/sql-createtype.html#id-1.9.3.94.5.8)**. For example:
```
CREATE EXTENSION uint;  -- add unsigned int datatypes / https://github.com/petere/pguint
```
([pguint](https://github.com/petere/pguint) is a popular extension, written by _PostgreSQL Core Team member_).
**There are 4 artifacts with user base types:**
1) If you restart DBeaver, it "forgets" about user types (in particular, types added by the pguint extension). User base types, user domain types and user enum types. Specifically, when creating a column with a custom type (or changing the column type to a custom type), BDeaver generates the "invalid data type" error. If you open the "Data Types" item in the left panel tree, then dbeaver sees the custom types and the error is not generated.
2) There are no custom types (base, domain, enum) in the data type selection list.
3) When sorting data by field of custom integer base types (for example, uint4) via the UI, DBeaver sorts values as strings, not as numbers.
4) Custom integer base types (for example, uint4) view as strings, not as numbers (black color instead blue and without space between triads).

This ussue is a continuation of the [bug  #10353](https://github.com/dbeaver/dbeaver/issues/10353)

**Comment by kseniiaguzeeva** [comment]
It should be fixed in the latest version https://dbeaver.io/files/ea/.
Please ask to reopen if it is still actual for you.

**Comment by vicreal** [comment]
DBeaver **7.3.3**.202101130717. **All of the problems have been**:

![0](https://user-images.githubusercontent.com/72701351/104468361-6a4a3380-55c8-11eb-8e49-a7bea28d502b.jpg)

> 
**1**. If you restart DBeaver, it "forgets" about user types (in particular, types added by the pguint extension). User base types, user domain types and user enum types. Specifically, when creating a column with a custom type (or changing the column type to a custom type), BDeaver generates the "invalid data type" error. If you open the "Data Types" item in the left panel tree, then dbeaver sees the custom types and the error is not generated.

For example, enter data type **uint4** and then press Enter or Tab - getting an error:
![Bug 1](https://user-images.githubusercontent.com/72701351/104467327-3fabab00-55c7-11eb-929e-4f3b86bde489.jpg)


**2**. There are no custom types (base, domain, enum) in the data type selection list.
**uint1**, **unit2**, **unit4**, **unit8**, **int1** absent in list:
![Bug 2](https://user-images.githubusercontent.com/72701351/104467334-42a69b80-55c7-11eb-81ec-eb03681647b2.jpg)

**3**. When sorting data by field of custom integer base types (for example, uint4) via the UI, DBeaver sorts values as strings, not as numbers.
```
CREATE TABLE public.test (num uint4 NOT NULL);
INSERT INTO public.test(num) VALUES(1),(2),(3),(4),(5),(6),(7),(8),(9),(10),(11),(12),(13),(14),(15),(16),(17),(18),(19),(20),(21),(22);
SELECT * FROM public.test;
```
Then sort column via the UI (click on the arrow):
![Bug 3](https://user-images.githubusercontent.com/72701351/104467411-5a7e1f80-55c7-11eb-96b2-8f28b5f20346.jpg)

**4**. Custom integer base types (for example, uint4) view as strings, not as numbers (black color instead blue and without space between triads).
`SELECT 12567::int4, 12567::uint4;`

![Bug 4](https://user-images.githubusercontent.com/72701351/104467782-c52f5b00-55c7-11eb-94d9-12b71ac0a5c2.jpg)

**Comment by LonwoLonwo** [comment]
@vicreal Unfortunately, PostgreSQL is a kind of database that can have a lot of data types. Previously, we read them all for all schemes, but this led to a slowdown in processes. 
Therefore, we began to load only the main list of types and load as needed. 
As a workaround now - on initial, go to the public schema and open the "Data Types" tab. Then access to custom types will be in all schemes, in theory, because they are cached.

We are working on increasing the list of loaded data types, most likely, this will be changed in the next release.

**Comment by LonwoLonwo** [comment]
And about point 3 and 4. We read data type information from pg_class.pg_type table. As I can see, rows for uint4, uint8 do not contain any hints that this custom type is numeric. For example, type "int8" has letter "N" in "typcategory", which tells us that he is numeric. But all custom types have letter "U" in this column. 

Maybe you have any idea how to determine that a custom type is numeric?

**Comment by vicreal** [comment]
> And about point 3 and 4. We read data type information from pg_class.pg_type table. As I can see, rows for uint4, uint8 do not contain any hints that this custom type is numeric. For example, type "int8" has letter "N" in "typcategory", which tells us that he is numeric. But all custom types have letter "U" in this column.
> Maybe you have any idea how to determine that a custom type is numeric?

Apparently, by "_pg_class_" you meant "_pg_type_". Yes, in general (only on the basis of this table), the numeric category of the type cannot be determined, because the "typcategory" field is "U", and on the basis of other fields, the data type does not belong to numeric types.

I can offer the following options:
**1)** Binding to the type name: `"pg_type.typname" IN ('int1', 'uint1', 'uint2', 'uint4', 'uint8')`
**2)** Binding to the package name: `pg_proc.probin LIKE '%/uint' WHERE (pg_type.typinput = pg_proc.proname)`
**3)** Surely there is still some clever way to determine that these are numeric types ))

**Comment by vicreal** [comment]
> @vicreal Unfortunately, PostgreSQL is a kind of database that can have a lot of data types. Previously, we read them all for all schemes, but this led to a slowdown in processes.
> Therefore, we began to load only the main list of types and load as needed.
> As a workaround now - on initial, go to the public schema and open the "Data Types" tab. Then access to custom types will be in all schemes, in theory, because they are cached.
> 
> We are working on increasing the list of loaded data types, most likely, this will be changed in the next release.

Possible variants:
**1)** load all data types from all schemas once when loading DBeaver
**2)** add an option in the settings "load only basic types" or "load all types from all schemas" when loading DBeaver
**3)** load all data types from the current schema (if not already loaded) at the moment when a table field is creating or changing, as well as in other cases when you need to enter a data type or select a data type from the selection list.

**Comment by LonwoLonwo** [comment]
oh, yes my mistake.  pg_class = pg_catalog

**Comment by serge-rider** [comment]
Perhaps related to #6110

**Comment by LonwoLonwo** [comment]
Hello @vicreal 
Can you test new version 7.3.5? Probably the first two points could be corrected.

**Comment by vicreal** [comment]
> Hello @vicreal
> Can you test new version 7.3.5? Probably the first two points could be corrected.

Tested **7.3.5** version on **uint** types: first and second problems solved.
Custom data types can be written or checked in the selection list (custom data types present in selection list).

Geomerty and geography types (from Postgis extension) is also [present](https://github.com/dbeaver/dbeaver/issues/6110) in selection list.

**Comment by LonwoLonwo** [comment]
Yep, thanks.

**Comment by LonwoLonwo** [comment]
Fixed in DBeaver 21.0.4
uint types + int1 type are numeric now (If they are in the data types list)

**Comment by uslss** [comment]
verified

**Comment by vicreal** [comment]
> _Fixed in DBeaver 21.0.4_
> _verified_

How can I download version 21.0.4 ?
[Hear](https://dbeaver.io/files/ea/) only 21.0.3...

**Comment by LonwoLonwo** [comment]
Hello @vicreal 
you can do it now from https://dbeaver.io/

**Comment by vicreal** [comment]
Tested version 21.0.4: **problen 3** was fixed (sorting as numbers, not as strings).
But **problem 4** still the case:

> 4. Custom integer base types (for example, uint4) view as strings, not as numbers (black color instead blue and without space between triads).
> SELECT 12567::int4, 12567::uint4;
> 
> ![Bug 4](https://user-images.githubusercontent.com/72701351/116940771-13e82a00-ac77-11eb-991e-47aa1f103946.jpg)

**Comment by LonwoLonwo** [comment]
Can't reproduce. 
![2021-05-04 01_29_55-DBeaver Enterprise 21 0 4 - _none_ Script-23](https://user-images.githubusercontent.com/45152336/116941436-578f6380-ac78-11eb-9cc9-e54f72ec7c84.png)

**Comment by vicreal** [comment]
Instead of bug 3, a new bug appeared: all values of uint types show as NULL
Example:
```
CREATE TABLE public.test (num uint4 NOT NULL);
INSERT INTO public.test(num) VALUES(1),(2),(3),(4),(5),(6),(7),(8),(9),(10),(11),(12),(13),(14),(15),(16),(17),(18),(19),(20),(21),(22);
SELECT * FROM public.test;
```
Then double click on table "test":
![Bug 3 new](https://user-images.githubusercontent.com/72701351/116942118-a5f13200-ac79-11eb-832d-00396ed0d56f.jpg)

`SELECT * FROM public.test WHERE (num = 2) OR (num = 3);`
![Bug 3 new 2](https://user-images.githubusercontent.com/72701351/116942209-dcc74800-ac79-11eb-93c5-25c56bac09e8.jpg)

![2021-05-04_014616](https://user-images.githubusercontent.com/72701351/116942613-98887780-ac7a-11eb-8068-fd87e70daad7.jpg)

**Comment by LonwoLonwo** [comment]
Sorry, but can't reproduce again.

![2021-05-04 01_52_28-DBeaver Enterprise 21 0 4 - _gis_ Script-24](https://user-images.githubusercontent.com/45152336/116943093-8955f980-ac7b-11eb-85be-cf6eafe5df2c.png)

**Comment by vicreal** [comment]
The chronology is as follows:
**1)** Closed DBeaver, updated 21.0.3 to 21.0.4, started DBeaver. Checked: bug 3 fixed, **bug 4 not fixed**.
**2)** Further found that in all tables the uint data types are displayed as **NULL**.
**3)** Rebooted Dbeaver. Checked: **bug 3 is not fixed**, **bug 4 is not fixed**, in all tables uint data types are displayed as **NULL**.
**4)** Opened the data types in the navigator - after that, all the problems disappeared: bug 3 fixed, bug 4 fixed, uint data types are displayed correctly.
**5)** Reloaded DBeaver again - everything works, there are no problems (without opening data types in the navigator).

**Conclusion**: after upgrading to version **21.0.4**, you need to open the data types in the navigator once. And only after that, there are guaranteed to be no problems.

**Comment by vicreal** [comment]
If it's not difficult, tell us what option you implemented to determine that **uint extension** types are **integer** types:

> 1) Binding to the type name: "pg_type.typname" IN ('int1', 'uint1', 'uint2', 'uint4', 'uint8')
> 2) Binding to the package name: pg_proc.probin LIKE '%/uint' WHERE (pg_type.typinput = pg_proc.proname)
> 3) Surely there is still some clever way to determine that these are numeric types ))

**Comment by LonwoLonwo** [comment]
> If it's not difficult, tell us what option you implemented to determine that uint extension types are integer types:

It was like 

> Binding to the type name: "pg_type.typname" IN ('int1', 'uint1', 'uint2', 'uint4', 'uint8')

But from Java side

**Comment by vicreal** [comment]
> 
> 
> The chronology is as follows:
> **1)** Closed DBeaver, updated 21.0.3 to 21.0.4, started DBeaver. Checked: bug 3 fixed, **bug 4 not fixed**.
> **2)** Further found that in all tables the uint data types are displayed as **NULL**.
> **3)** Rebooted Dbeaver. Checked: **bug 3 is not fixed**, **bug 4 is not fixed**, in all tables uint data types are displayed as **NULL**.
> **4)** Opened the data types in the navigator - after that, all the problems disappeared: bug 3 fixed, bug 4 fixed, uint data types are displayed correctly.
> **5)** Reloaded DBeaver again - everything works, there are no problems (without opening data types in the navigator).
> 
> **Conclusion**: after upgrading to version **21.0.4**, you need to open the data types in the navigator once. And only after that, there are guaranteed to be no problems.


No, this is a **stable bug**. Reproducible easily (version **21.0.4**):
**1)** Open DBeaver
**2)** In the left panel "Databases" click on the "public" schema and press **F5** => DBeaver updates tables in public schema
**3)** After that DBeaver forgets user data types => **bugs 1,2,3,4** appear again and **all uint types show as NULL** (see my screenshots above):

![Bug 3 and 4 again](https://user-images.githubusercontent.com/72701351/118380291-4552e500-b5e9-11eb-8d73-bc14f58f51f4.jpg)

**Comment by LonwoLonwo** [comment]
Ok, now I can reproduce it. Thanks for investigation.

**Comment by LonwoLonwo** [comment]
> 1) Open DBeaver
> 2) In the left panel "Databases" click on the "public" schema and press F5 => DBeaver updates tables in public schema
> 3) After that DBeaver forgets user data types => bugs 1,2,3,4 appear again and all uint types show as NULL

Fixed in DBeaver 21.1.0

**Comment by kseniiaguzeeva** [comment]
verified
